### DISCUSSION TITLE
Recommended user onboarding and subscription flow

### DISCUSSION THREAD

**Discussion by clayton** [pay-rails/pay#1066] [Q&A] [answered]
I am integrating Pay into my SaaS application and wondering if there's a recommended flow for onboarding users and setting up subscriptions.

My use case is fairly common:

- Users can create accounts without subscribing (free tier)
- Users can upgrade to a paid account by subscribing (customers)
- I'm using Stripe w/ Stripe Checkout

Using a pricing table (and Stripe Checkout) I am able to get Pay Customer records created with the web hooks, but not really sure where I should be creating the subscriptions and/or in what order I should be creating Pay Customers, syncing subscriptions, or creating subscriptions.

Overall I'm confused about how much I need to be explicitly creating Pay records vs. not.

Things like:

- Do I need to manually create subscriptions using `@user.payment_processor.subscribe(name: "default", plan: "monthly")`?
- Should I create a customer record manually `@user.payment_processor.customer` or rely on the web hooks to do that for me?
- Should I be redirecting my user's to a custom page after checkout to handle subscription creation or try to let the web hooks handle that?
- Why do I keep getting duplicate stripe customer records with the same email address?
- Why are there never records populated in the Pay::Subscription table after a user completes the stripe checkout flow?

Would love to get a handle on a recommended approach and then contribute something back to the docs.

**Comment by excid3** [comment, marked as answer]
You have to associate things with the Stripe Customer ID in the database so Pay can sync webhooks for the customer. If you're not providing a Stripe Customer ID to the pricing table, it will create a new customer each time. Pay will see the webhook but not a matching customer so it will ignore it.

Recommend reading Stripe's documentation for how things work: https://docs.stripe.com/payments/checkout/pricing-table#customer-session

I just added a `customer_session` method to handle this for you:

```ruby
class PricingController < ApplicationController 
  def show
    # sets Stripe as the payment processor which lazily creates a Stripe Customer when you use it
    current_user.set_payment_processor(:stripe)

    # Ensures a Stripe Customer exists, and creates a CustomerSession object.
    @customer_session = current_user.payment_processor.customer_session(components: {
      pricing_table: {enabled: true}
    }) 
  end
end
```

```erb
  <script
    async
    src="https://js.stripe.com/v3/pricing-table.js">
  </script>
  <stripe-pricing-table
    pricing-table-id='{{PRICING_TABLE_ID}}'
    publishable-key="pk_test_1234"
    customer-session-client-secret="<%= @customer_session.client_secret %>"
  >
  </stripe-pricing-table>
```

**Comment by clayton** [comment]
Thanks, @excid3! This makes more sense now.

I had only got as far as adding a `client-reference-id` to the pricing table (the user's id) but didn't get to the `customer-session-client-secret` part in the stripe docs.

I'll go implement this and come back with a step-by-step for future reference.
